src/include/migraphx/op/dot_broadcast.hpp

```diff
@@ -0,0 +1,94 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2015-2023 Advanced Micro Devices, Inc. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#ifndef MIGRAPHX_GUARD_OPERATORS_DOT_BROADCAST_HPP
+#define MIGRAPHX_GUARD_OPERATORS_DOT_BROADCAST_HPP
+
+#include <migraphx/check_shapes.hpp>
+#include <migraphx/config.hpp>
+#include <migraphx/common.hpp>
+#include <migraphx/dyn_output.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+namespace op {
+
+/**
+ * Broadcast dimensions between two tensors for the `dot` operator.
+ * Essentially broadcasts between two shapes for dimensions other than the last two.
+ * This operator is only needed if one of the shapes are dynamic.
+ * Example:
+ * a = shape[{1, 4}, 3, 248, 248]
+ * b = shape[248, 365]
+ * dot_broadcast(a, b) => shape[{1, 4}, 3, 248, 248] (no change)
+ * dot_broadcast(b, a) => shape[{1, 4}, 3, 248, 365]
+ */
+struct dot_broadcast
+{
+    std::string name() const { return "dot_broadcast"; }
```

Comment:
Maybe we should call this `broadcast_for_dot`. I thought at first this was doing a dot and broadcast together.